File: solver/solver.py
from collections import deque
from ..games.game import *
from ..database.database import GameDB

REMOTENESS_TERMINAL = 0

class Solver:
    def __init__(self, game: Game):
        self.game = game
        self.solution = {}
        self.parent_map = {}

    def solve(self, overwrite=False):
        self.db = GameDB(self.game.id)
        if overwrite or not self.db.exists:
            self.db.create_table(overwrite)
            self.discover()
            print("discovered")
            self.propagate()
            print("solved")
            self.db.insert(self.solution)

    def get_children(self, position):
        moves = self.game.generate_moves(position)
        return map(lambda m: self.game.do_move(position, m), moves)

    def discover(self):
        visited = set()
        q = deque()
        start = self.game.start()
        q.appendleft(start)
        visited.add(start)
        while q:
            position = q.pop()
            value = self.game.primitive(position)
            if value is not None:
                self.solution[position] = (REMOTENESS_TERMINAL, value)
            else:
                children = self.get_children(position)
                for child in children:
                    if not self.parent_map.get(child):
                        self.parent_map[child] = set()
                    self.parent_map[child].add(position)
                    if child not in visited:
                        visited.add(child)
                        q.appendleft(child)
    
    def propagate(self):
        q = deque(self.solution.keys())
        while q:
            position = q.pop()
            parent_rem, val = self.solution.get(position)
            parent_rem += 1
            parent_val = self.parent_value(val)
            parents = self.parent_map.get(position, set())
            for parent in parents:
                parent_sol = self.solution.get(parent)
                if parent_sol is not None:
                    self.solution[parent] = (parent_rem, parent_val)
                    q.appendleft(parent)
                elif parent_sol[1] < parent_val:
                    self.solution[parent] = (parent_rem, parent_val)

    
    def parent_value(self, val: Value) -> Value:
        if self.game.n_players == 1:
            return val
        else:
            if val == Value.Win:
                return Value.Loss
            elif val == Value.Loss:
                return Value.Win
            else:
                return val
    
    def print(self):
        if self.solution:
            sol = [(position, rem, value) for position, (rem, value) in self.solution.items()]
        else:
            sol = self.db.get_all()
        for (position, rem, value) in sol:
            print(f'state: {position} | remoteness: {rem} | value: {Value(value).name}')
    
    def get_remoteness(self, state: int) -> int:
        rem, _ = self.db.get(state)
        if rem is None:
            return -1
        return rem

File: solver/cppsolver/solver.cpp
#include "solver.hpp"
#include <queue>

using std::queue;

Solver::Solver() {
    Py_Initialize();
    do_move = PyObject_GetAttrString(module, "call_do_move");
}

void Solver::solve(const string& puzzle_id) {
    discover();
    propagate();
}

vector<State> Solver::getChildren(State s) const {

}

void Solver::discover() {
    State start;
    unordered_set<State> visited;
    queue<State> q;
    q.push(start);
    visited.insert(start);
    while (!q.empty()) {
        State position = q.front();
        q.pop();
        uint8_t value;
        if (value == 0) {
            rem_map.emplace(position, 0xff);
        } else if (value == 1) {
            rem_map.emplace(position, 0);
            primitives.insert(position);
        } else {
            vector<State> children = getChildren(position);
            for (State child : children) {
                if (!parent_map.contains(child)) {
                    parent_map.emplace(child, unordered_set<State>());
                }
                parent_map[child].insert(position);
                if (!visited.contains(child)) {
                    visited.insert(child);
                    q.push(child);
                }
            }
        }

    }
}

void Solver::propagate() {
    queue<State> q;
    for (State s : primitives) {
        q.push(s);
    }
    while (!q.empty()) {
        State position = q.front();
        q.pop();
        Remoteness rem = rem_map.at(position) + 1;
        unordered_set<State> parents = parent_map.at(position);
        for (State parent : parents) {
            if (!rem_map.contains(parent)) {
                rem_map.emplace(parent, rem);
                q.push(parent);
            }
        }
    }
}